arch/x86/core/x86_mmu.c

```diff
@@ -784,393 +1151,472 @@ int arch_buffer_validate(void *addr, size_t size, int write)
 {
 	struct k_mem_domain *domain = thread->mem_domain_info.mem_domain;
 
-	/* Non-user threads don't have per-thread page tables set up */
 	if ((thread->base.user_options & K_USER) == 0) {
 		return;
 	}
 
-	for (int i = 0, pcount = 0; pcount < domain->num_partitions; i++) {
-		struct k_mem_partition *partition;
-
-		partition = &domain->partitions[i];
-		if (partition->size == 0) {
-			continue;
-		}
-		pcount++;
-
-		reset_mem_partition(thread, partition);
+	if ((thread->base.thread_state & _THREAD_DEAD) == 0) {
+		/* Thread is migrating to another memory domain and not
+		 * exiting for good; we weren't called from
+		 * z_thread_single_abort().  Resetting the stack region will
+		 * take place in the forthcoming thread_add() call.
+		 */
+		return;
 	}
+
+	/* Restore permissions on the thread's stack area since it is no
+	 * longer a member of the domain.
+	 */
+	reset_region(domain->arch.ptables, (void *)thread->stack_info.start,
+		     thread->stack_info.size);
 }
 
 void arch_mem_domain_partition_add(struct k_mem_domain *domain,
 				   uint32_t partition_id)
 {
-	sys_dnode_t *node, *next_node;
-
-	SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
-		struct k_thread *thread =
-			CONTAINER_OF(node, struct k_thread, mem_domain_info);
-
-		if ((thread->base.user_options & K_USER) == 0) {
-			continue;
-		}
+	struct k_mem_partition *partition = &domain->partitions[partition_id];
 
-		apply_mem_partition(thread, &domain->partitions[partition_id]);
-	}
+	/* Update the page tables with the partition info */
+	apply_region(domain->arch.ptables, (void *)partition->start,
+		     partition->size, partition->attr | MMU_P);
 }
 
+/* Invoked from memory domain API calls, as well as during thread creation */
 void arch_mem_domain_thread_add(struct k_thread *thread)
 {
-	if ((thread->base.user_options & K_USER) == 0) {
-		return;
+	/* New memory domain we are being added to */
+	struct k_mem_domain *domain = thread->mem_domain_info.mem_domain;
+	/* This is only set for threads that were migrating from some other
+	 * memory domain; new threads this is NULL
+	 */
+	pentry_t *old_ptables = (pentry_t *)thread->arch.ptables;
+	bool is_user = (thread->base.user_options & K_USER) != 0;
+	bool is_migration = (old_ptables != NULL) && is_user;
+			     ;
+	/* Allow US access to the thread's stack in its new domain if
+	 * we are migrating. If we are not migrating this is done in
+	 * z_x86_current_stack_perms()
+	 */
+	if (is_migration) {
+		set_stack_perms(thread, domain->arch.ptables);
 	}
 
-	z_x86_apply_mem_domain(thread, thread->mem_domain_info.mem_domain);
+	thread->arch.ptables = (uintptr_t)domain->arch.ptables;
+	LOG_DBG("set thread %p page tables to %p", thread,
+		(void *)thread->arch.ptables);
+
+	/* Check if we're doing a migration from a different memory domain
+	 * and have to remove permissions from its old domain.
+	 *
+	 * XXX: The checks we have to do here and in
```

Comment:
Enhancement link for these `is_migration` hacks: https://github.com/zephyrproject-rtos/zephyr/issues/29601